File: services/agent-runtime/src/axiom/canonicalise.py
"""Canonical JSON + SHA-256 helpers.

Mirrors the TypeScript implementation in packages/ledger/src/canonicalise.ts.
Used for the hash-chained audit ledger and approval token payloads.
"""

from __future__ import annotations

import hashlib
import json
from typing import Any


def canonical_json(value: Any) -> str:
    """Produce a deterministic JSON string for any JSON-compatible value.

    - Object keys are sorted lexicographically
    - Arrays preserve order
    - Numbers/strings/booleans/null serialise natively
    - None is treated as null
    - No whitespace
    - UTF-8 encoded

    Mirrors packages/ledger/src/canonicalise.ts so the TS-side ledger
    and Python-side agents produce identical hashes for the same input.
    """
    return json.dumps(_sort(value), separators=(",", ":"), ensure_ascii=False, allow_nan=False)


def _sort(value: Any) -> Any:
    if value is None:
        return None
    if isinstance(value, (str, int, float, bool)):
        return value
    if isinstance(value, list):
        return [_sort(v) for v in value]
    if isinstance(value, dict):
        return {k: _sort(v) for k, v in sorted(value.items())}
    raise TypeError(f"Cannot canonicalise value of type {type(value)}")


def sha256_hex(data: str | bytes) -> str:
    """Return the lowercase hex SHA-256 digest of the input."""
    if isinstance(data, str):
        data = data.encode("utf-8")
    return hashlib.sha256(data).hexdigest()

File: packages/ledger/src/canonicalise.ts
/**
 * Canonical JSON serialisation.
 *
 * Why not just JSON.stringify? Because two semantically identical objects
 * can serialise to different strings depending on key order, which would
 * break the hash chain. This module produces a canonical, deterministic
 * string for any JSON-compatible value.
 *
 * Rules:
 *   - Object keys are sorted lexicographically
 *   - Arrays preserve order
 *   - Numbers, strings, booleans, null serialise natively
 *   - undefined is treated as null (matching JSON)
 *   - No whitespace
 *   - UTF-8 encoded
 */

type JsonValue =
  string | number | boolean | null | undefined | JsonValue[] | { [k: string]: JsonValue };

function sortKeys(value: JsonValue): JsonValue {
  if (value === null || value === undefined) return null;
  if (Array.isArray(value)) {
    return value.map(sortKeys);
  }
  if (typeof value === 'object') {
    const sorted: { [k: string]: JsonValue } = {};
    for (const key of Object.keys(value as Record<string, JsonValue>).sort()) {
      const v = (value as Record<string, JsonValue>)[key];
      if (v === undefined) {
        sorted[key] = null;
      } else {
        sorted[key] = sortKeys(v);
      }
    }
    return sorted;
  }
  return value;
}

export function canonicalJson(value: unknown): string {
  return JSON.stringify(sortKeys(value as JsonValue));
}

export async function sha256(input: string | Uint8Array): Promise<string> {
  // Copy byte inputs into an ArrayBuffer-backed typed array. WebCrypto rejects
  // views that could be backed by a SharedArrayBuffer, and TS 6 reflects that.
  const data = typeof input === 'string' ? new TextEncoder().encode(input) : new Uint8Array(input);
  const hash = await crypto.subtle.digest('SHA-256', data);
  return Array.from(new Uint8Array(hash))
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
}
